spring-core/src/test/java/org/springframework/util/MultiValueMapRelatedTests.java

```diff
@@ -0,0 +1,214 @@
+	}
+
+	@ParameterizedTest
+	@MethodSource("objectsUnderTest")
+	void toSingleValueMap(MultiValueMap<String, String> objectUnderTest) {
+
+		int startingSize = objectUnderTest.size();
+
+		List<String> values = new ArrayList<>(2);
+		values.add("value1");
+		values.add("value2");
+		objectUnderTest.put("key", values);
+		Map<String, String> singleValueMap = objectUnderTest.toSingleValueMap();
+		assertThat(singleValueMap).hasSize(startingSize + 1);
+		assertThat(singleValueMap.get("key")).isEqualTo("value1");
+	}
+
+	@ParameterizedTest
+	@MethodSource("objectsUnderTest")
+	@Disabled("to be fixed in gh-25140")
+	void toSingleValueMapWithEmptyList(MultiValueMap<String, String> objectUnderTest) {
+		objectUnderTest.put("key", Collections.emptyList());
+		Map<String, String> singleValueMap = objectUnderTest.toSingleValueMap();
+		assertThat(singleValueMap).isEmpty();
+		assertThat(singleValueMap.get("key")).isNull();
+	}
+
+	@ParameterizedTest
+	@MethodSource("objectsUnderTest")
+	void equalsOnExistingValues(MultiValueMap<String, String> objectUnderTest) {
+		objectUnderTest.clear();
+		objectUnderTest.set("key1", "value1");
+		assertThat(objectUnderTest).isEqualTo(objectUnderTest);
+	}
+
+	@ParameterizedTest
+	@MethodSource("objectsUnderTest")
+	void equalsOnEmpty(MultiValueMap<String, String> objectUnderTest) {
+		objectUnderTest.clear();
+		objectUnderTest.set("key1", "value1");
+		MultiValueMap<String, String> o1 = new LinkedMultiValueMap<>();
+		o1.set("key1", "value1");
+		assertThat(o1).isEqualTo(objectUnderTest);
+		assertThat(objectUnderTest).isEqualTo(o1);
+		Map<String, List<String>> o2 = new HashMap<>();
+		o2.put("key1", Collections.singletonList("value1"));
+		assertThat(o2).isEqualTo(objectUnderTest);
+		assertThat(objectUnderTest).isEqualTo(o2);
+	}
+
+	@ParameterizedTest
+	@MethodSource("objectsUnderTest")
+	void canNotChangeAnUnmodifiableMultiValueMap(MultiValueMap<String, String> objectUnderTest) {
+		MultiValueMap<String, String> asUnmodifiableMultiValueMap = CollectionUtils.unmodifiableMultiValueMap(objectUnderTest);
+		Assertions.assertAll(
+				() -> Assertions.assertThrows(UnsupportedOperationException.class,
+						() -> asUnmodifiableMultiValueMap.add("key", "value")),
+				() -> Assertions.assertThrows(UnsupportedOperationException.class,
+						() -> asUnmodifiableMultiValueMap.addIfAbsent("key", "value")),
+				() -> Assertions.assertThrows(UnsupportedOperationException.class,
+						() -> asUnmodifiableMultiValueMap.addAll("key", exampleListOfValues())),
+				() -> Assertions.assertThrows(UnsupportedOperationException.class,
+						() -> asUnmodifiableMultiValueMap.addAll(exampleMultiValueMap())),
+				() -> Assertions.assertThrows(UnsupportedOperationException.class,
+						() -> asUnmodifiableMultiValueMap.set("key", "value")),
+				() -> Assertions.assertThrows(UnsupportedOperationException.class,
+						() -> asUnmodifiableMultiValueMap.setAll(exampleHashMap())),
```

Comment:
We tend to use `Collections.singletonMap` for such use cases.

```suggestion
						() -> asUnmodifiableMultiValueMap.setAll(Collections.singletonMap("key2", "key2.value1")),
```